// library/BtnTransparent.js
/* 
		WARNING 
		Auto generated file. 
		Do not modify its contents.
*/

const extend = require('js-base/core/extend');
const Button = require('sf-core/ui/button');
const Color = require('sf-core/ui/color');
const FlexLayout = require('sf-core/ui/flexlayout');



const getCombinedStyle = require("library/styler-builder").getCombinedStyle;

const BtnTransparent = extend(FlexLayout)(
	//constructor
	function(_super, props) {
		// initalizes super class for this component scope
		_super(this, props);

		const button1Style = getCombinedStyle(".button", {
			width: null,
			height: null,
			borderRadius: 22,
			backgroundColor: Color.create(0, 0, 161, 241),
			borderWidth: 1,
			borderColor: Color.create(77, 255, 255, 255),
			flexGrow: 1
		});
		var button1 = new Button(button1Style);
		this.addChild(button1);
		this.button1 = button1;

		//assign the children to btnTransparent 
		this.children = Object.assign({}, {
			button1: button1
		});
		
	});

BtnTransparent.defaults = getCombinedStyle(".flexLayout", {
	backgroundColor: Color.create(0, 255, 255, 255),
	width: 180,
	height: 45,
	borderColor: Color.create(77, 255, 255, 255),
	borderWidth: null
});

module && (module.exports = BtnTransparent);
