// library/Flexlayout3.js
/* 
		WARNING 
		Auto generated file. 
		Do not modify its contents.
*/

const extend = require('js-base/core/extend');
const FlexLayout = require('sf-core/ui/flexlayout');
const Color = require('sf-core/ui/color');

const ItemSmallThumb = require("../../components/ItemSmallThumb");

const getCombinedStyle = require("library/styler-builder").getCombinedStyle;

const Flexlayout3 = extend(FlexLayout)(
	//constructor
	function(_super, props) {
		// initalizes super class for this component scope
		_super(this, props);

		const flexlayout4Style = getCombinedStyle(".flexLayout .flexLayout-smallThumb.inactive", {
			left: 0,
			top: 0,
			width: 60,
			height: 60,
			borderColor: Color.create(26, 0, 0, 0),
			positionType: FlexLayout.PositionType.RELATIVE
		});
		var flexlayout4 = new ItemSmallThumb(flexlayout4Style, "flexlayout3");
		this.addChild(flexlayout4);
		
		//assign the children to flexlayout3 
		this.children = Object.assign({}, {
			flexlayout4: flexlayout4
		});
		
	});

Flexlayout3.defaults = getCombinedStyle(".flexLayout", {
	width: 354,
	height: 70,
	backgroundColor: Color.create(255, 255, 255, 255),
	flexDirection: FlexLayout.FlexDirection.ROW,
	justifyContent: FlexLayout.JustifyContent.FLEX_START,
	alignContent: FlexLayout.AlignContent.CENTER,
	alignItems: FlexLayout.AlignItems.CENTER
});

module && (module.exports = Flexlayout3);
