// library/LoaderContainer.js
/* 
		WARNING 
		Auto generated file. 
		Do not modify its contents.
*/

const extend = require('js-base/core/extend');
const ActivityIndicator = require('sf-core/ui/activityindicator');
const Color = require('sf-core/ui/color');
const FlexLayout = require('sf-core/ui/flexlayout');



const getCombinedStyle = require("library/styler-builder").getCombinedStyle;

const LoaderContainer = extend(FlexLayout)(
	//constructor
	function(_super, props) {
		// initalizes super class for this component scope
		_super(this, props);

		const activityIndicatorStyle = getCombinedStyle(".activityIndicator", {
			backgroundColor: Color.create(0, 255, 255, 255),
			id: 1005,
			color: Color.create(255, 211, 210, 210),
			minHeight: 21,
			minWidth: 21
		});
		var activityIndicator = new ActivityIndicator(activityIndicatorStyle);
		this.addChild(activityIndicator);
		this.activityIndicator = activityIndicator;

		//assign the children to loaderContainer 
		this.children = Object.assign({}, {
			activityIndicator: activityIndicator
		});
		
	});

LoaderContainer.defaults = getCombinedStyle(".flexLayout", {
	width: 102,
	height: 104,
	backgroundColor: Color.create(0, 255, 255, 255),
	id: 1004,
	alignItems: FlexLayout.AlignItems.CENTER,
	justifyContent: FlexLayout.JustifyContent.CENTER
});

module && (module.exports = LoaderContainer);
